osbuild/pipeline.py

```diff
@@ -210,82 +210,112 @@ def description(self):
             description["assembler"] = self.assembler.description()
         return description
 
-    @contextlib.contextmanager
-    def get_buildtree(self, object_store):
-        if self.build:
-            with object_store.new() as tree:
-                tree.base = self.build.tree_id
-                with tree.read() as path:
-                    yield path
-        else:
-            with object_store.tempdir() as tmp:
-                objectstore.mount("/", tmp)
-                try:
-                    yield tmp
-                finally:
-                    objectstore.umount(tmp)
+    def build_stages(self, object_store, interactive, libdir, secrets):
+        results = {"success": True}
 
-    def run(self, store, interactive=False, libdir=None, secrets=None):
-        os.makedirs("/run/osbuild", exist_ok=True)
-        object_store = objectstore.ObjectStore(store)
-        results = {}
+        if not self.stages:
+            return results
 
-        if self.build and self.build.stages:
-            # For now, the last build stage is always committed to the object store
-            self.build.stages[-1].checkpoint = True
+        # We need a build tree for the stages below, which is either
+        # another tree that needs to be built with the build pipeline
+        # or the host file system if no build pipeline is specified
+        # NB: the very last level of nested build pipelines is always
+        # build on the host
+
+        if not self.build:
+            build_tree = objectstore.HostTree(object_store)
+        else:
+            build = self.build
+            assert build.stages
```

Comment:
Does this mean a build pipeline with `"stages": []` fails? Not that that is important, but it seem awkward.

Reply:
I assumed that this was a malformed manifest, i.e. having a build pipeline section but not stages, but I could well be missing something.

Reply:
It is not malformed according to [the schema](https://github.com/osbuild/osbuild/blob/27496fa10f6a7091be8710d828da40ca05670508/schemas/osbuild1.json) @dvdhrm wrote down. That doesn't necessarily need to be correct, but we should make sure we're consistent.

This check is only for the build pipeline, not for the normal pipeline, right?

Reply:
I removed the check because it should not cause any harm. It will just add another level of nested `"build": {"success": true}` object to the result json.